### PR TITLE
Git integration inside the IDE (commit only)

### PR DESCRIPTION
**Pull request by baydindima**
This add opportunity for commit with one-line message. New item "Commit" locate in Edit menu (I do not know how it wisely). I saw this issue https://github.com/arduino/Arduino/issues/4387 and decided to implement this. 
Please review and comment my solution.

### CODE REVIEW

**Comment by cmaglie** [comment]
@ArduinoBot build this please

File: app/src/cc/arduino/view/git/GitCommit.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  private javax.swing.JPanel buttonsContainer;
+  private javax.swing.JButton cancelButton;
+  private javax.swing.JTextField commitMessageField;
+  private javax.swing.JButton okButton;
+  // End of variables declaration//GEN-END:variables
+
+  private String[] getArgsForProcess(String command) {
+    String[] args = new String[3];
+    if (OSUtils.isWindows()) {
+      args[0] = "cmd.exe";
```
Comment (cmaglie) [inline app/src/cc/arduino/view/git/GitCommit.java]: Why not call directly git?

File: app/src/cc/arduino/view/git/GitCommit.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    args[2] = command;
+
+    return args;
+  }
+
+  private void setSystemEnvironment(Map<String, String> environment) {
+    environment.put("HOME", System.getProperty("user.home"));
+    environment.put("LANG", "en_US.UTF-8");
+    environment.put("GDM_LANG", "en_US.UTF-8");
+    environment.put("LANGUAGE", "us");
```
Comment (cmaglie) [inline app/src/cc/arduino/view/git/GitCommit.java]: AFAICS you don't parse the output of git, so why you set the locale to en_US?

File: app/src/cc/arduino/view/git/GitCommit.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  }
+
+  private void commit() {
+    String filePath = editor.getSketch().getMainFilePath();
+    StringBuilder command = new StringBuilder(String.format("git add %s && git commit %s ", filePath, filePath));
+
+    String message = commitMessageField.getText();
+    if (!message.isEmpty()) {
+      command.append(" -m \"")
+        .append(message)
```
Comment (cmaglie) [inline app/src/cc/arduino/view/git/GitCommit.java]: This doesn't handle correctly commit messages that contains "quotes"

**Comment by cmaglie** [comment]
Ok, it seems that @ArduinoBot is on holidays, so I've tested the patch manually.

First, some comments about the patch (I've added some other minor notes inline in the commits too)
- IMHO the "Commit" command should not appear in the menu if the "git" command is not available on the system path. As it is now you get a "file not found" error if you try to use it without git installed, but it's better to prevent the error message and not show the "Commit" command at all to the user if git is not available.
- In case "git" is available the "Commit" menu should be disabled (grayed out) in all the cases when the commit cannot be performed, for example, if the sketch folder is not a git repository.
- The patch doesn't handle sketch with multiple files (for example if you create a new tab with Ctrl+Shift+N the new file doesn't get committed, or if you remove a file the deletion is not committed)

Now, besides the notes above, It seems that the usecase is very narrow, becuase the user need anyway an external git client to "git clone" the sketch from a remote repository, or "git initialize" it after saving the sketch for the first time.
Also all the other operations (branching, push/pull, merge, etc..) needs an external client and seems improbable that all this being implemented in the Arduino IDE.
So I'm wondering if this patch is really worth the effort...

**Comment by NicoHood** [comment]
The bot is on holidays?? O_o

I'd like to test a build when its available. However from what I've heard it might make more sense to just add a command line to the IDE, rather than a "not so useful half complete" git implementation.

Because windows users wont use git (or only the bright and shiny github gui) and linux users just know how to use proper software the normal way.

**Pull request closed without merging**